tests/test_cargo_init.rs

```diff
@@ -0,0 +1,297 @@
+    
+    File::create(&sourcefile_path2).unwrap().write_all(br#"
+        fn main () {
+            println!("Hello, world 3!");
+        }
+    "#).unwrap();
+    
+    assert_that(cargo_process("init").arg("--vcs").arg("none")
+                                    .env("USER", "foo").cwd(&path),
+                execs().with_status(101));
+                
+    assert_that(&paths::root().join("foo/Cargo.toml"), is_not(existing_file()));
+    //assert_that(cargo_process("build").cwd(&paths::root().join("foo")),
+    //            execs().with_status(0));
+    
+});
+
+fn lib_already_exists(rellocation: &str) {
+    let path = paths::root().join("foo");
+    fs::create_dir(&path).ok();
+    fs::create_dir(&path.join("src")).ok();
+    
+    let sourcefile_path = path.join(rellocation);
+    
+    File::create(&sourcefile_path).unwrap().write_all(br#"
+        pub fn qqq() {}
+    "#).unwrap();
+    
+    assert_that(cargo_process("init").arg("--vcs").arg("none")
+                                    .env("USER", "foo").cwd(&path),
+                execs().with_status(0));
+
+    assert_that(&paths::root().join("foo/Cargo.toml"), existing_file());
+    assert_that(&paths::root().join("foo/src/main.rs"), is_not(existing_file()));
+    
+    // Check that our file is not overwritten
+    let mut contents = String::new();
+    File::open(&sourcefile_path).unwrap().read_to_string(&mut contents).unwrap();
+    assert!(contents.contains(r#"pub fn qqq() {}"#));
+
+    assert_that(cargo_process("build").cwd(&paths::root().join("foo")),
+                execs().with_status(0));
+}
+
+test!(lib_already_exists_src {
+    lib_already_exists("src/lib.rs")
+});
+
+test!(lib_already_exists_nosrc {
+    lib_already_exists("lib.rs")
+});
+
+test!(simple_git {
+    assert_that(cargo_process("init").arg("--vcs").arg("git")
+                                    .env("USER", "foo"),
+                execs().with_status(0));
+
+    assert_that(&paths::root().join("Cargo.toml"), existing_file());
+    assert_that(&paths::root().join("src/lib.rs"), existing_file());
+    assert_that(&paths::root().join(".git"), existing_dir());
+    assert_that(&paths::root().join(".gitignore"), existing_file());
+
+    assert_that(cargo_process("build").cwd(&paths::root().join(".")),
+                execs().with_status(0));
+});
+
+test!(git_autodetect {
+    fs::create_dir(&paths::root().join(".git")).ok();
+    
+    assert_that(cargo_process("init")
+                                    .env("USER", "foo"),
+                execs().with_status(0));
+    
+
+    assert_that(&paths::root().join("Cargo.toml"), existing_file());
+    assert_that(&paths::root().join("src/lib.rs"), existing_file());
+    assert_that(&paths::root().join(".git"), existing_dir());
+    assert_that(&paths::root().join(".gitignore"), existing_file());
+
+    assert_that(cargo_process("build").cwd(&paths::root().join(".")),
+                execs().with_status(0));
```

Comment:
It's ok to avoid most of the `cargo build` assertions at the end of all the tests here, that'll most likely end up just slowing down the tests without providing too many more assertions. I think it's important for one or two of the tests to test out `cargo build` after `cargo init`, but it's ok to not have that happen for all tests.